// ig-trading-api/dist/lightstreamer/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChartFields = exports.ChartResolution = void 0;
var ChartResolution;
(function (ChartResolution) {
    ChartResolution["HOUR"] = "HOUR";
    ChartResolution["MINUTE"] = "1MINUTE";
    ChartResolution["MINUTE_5"] = "5MINUTE";
    ChartResolution["SECOND"] = "SECOND";
})(ChartResolution = exports.ChartResolution || (exports.ChartResolution = {}));
var ChartFields;
(function (ChartFields) {
    ChartFields["BID_CLOSE"] = "BID_CLOSE";
    ChartFields["BID_HIGH"] = "BID_HIGH";
    ChartFields["BID_LOW"] = "BID_LOW";
    ChartFields["BID_OPEN"] = "BID_OPEN";
    ChartFields["CONS_END"] = "CONS_END";
    ChartFields["CONS_TICK_COUNT"] = "CONS_TICK_COUNT";
    ChartFields["DAY_HIGH"] = "DAY_HIGH";
    ChartFields["DAY_LOW"] = "DAY_LOW";
    ChartFields["DAY_NET_CHG_MID"] = "DAY_NET_CHG_MID";
    ChartFields["DAY_OPEN_MID"] = "DAY_OPEN_MID";
    ChartFields["DAY_PERC_CHG_MID"] = "DAY_PERC_CHG_MID";
    ChartFields["LTP_CLOSE"] = "LTP_CLOSE";
    ChartFields["LTP_HIGH"] = "LTP_HIGH";
    ChartFields["LTP_LOW"] = "LTP_LOW";
    ChartFields["LTP_OPEN"] = "LTP_OPEN";
    ChartFields["LTV"] = "LTV";
    ChartFields["OFR_CLOSE"] = "OFR_CLOSE";
    ChartFields["OFR_HIGH"] = "OFR_HIGH";
    ChartFields["OFR_LOW"] = "OFR_LOW";
    ChartFields["OFR_OPEN"] = "OFR_OPEN";
    ChartFields["TTV"] = "TTV";
    ChartFields["UTM"] = "UTM";
})(ChartFields = exports.ChartFields || (exports.ChartFields = {}));
