// ig-trading-api/dist/login/LoginAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoginAPI = void 0;
const axios_1 = __importDefault(require("axios"));
const APIClient_1 = require("../APIClient");
class LoginAPI {
    constructor(apiClient, auth) {
        this.apiClient = apiClient;
        this.auth = auth;
    }
    /**
     * Creates a trading session, obtaining session tokens for subsequent API access. Please note that region-specific
     * login restrictions may apply.
     *
     * @param username - Username
     * @param password - Password
     * @see https://labs.ig.com/rest-trading-api-reference/service-detail?id=534
     */
    createSession(username, password) {
        return __awaiter(this, void 0, void 0, function* () {
            delete this.auth.accessToken;
            const resource = LoginAPI.URL.SESSION;
            const response = yield this.apiClient.post(resource, {
                identifier: this.auth.username || username,
                password: this.auth.password || password,
            }, {
                'axios-retry': {
                    retries: 0,
                },
                headers: {
                    VERSION: '3',
                },
            });
            this.auth.accessToken = response.data.oauthToken.access_token;
            this.auth.refreshToken = response.data.oauthToken.refresh_token;
            this.auth.accountId = response.data.accountId;
            this.auth.lightstreamerEndpoint = response.data.lightstreamerEndpoint;
            yield this.getSessionToken();
            return response.data;
        });
    }
    /**
     * Switches active accounts, optionally setting the default IG account (of type CFD or spreadbet), against which trades may be made.
     *
     * @param accountId - Account ID
     * @see https://labs.ig.com/rest-trading-api-reference/service-detail?id=534
     */
    switchAccount(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = LoginAPI.URL.SESSION;
            const response = yield this.apiClient.put(resource, {
                accountId: accountId,
                defaultAccount: false,
            }, {
                'axios-retry': {
                    retries: 0,
                },
                headers: {
                    VERSION: '1',
                },
            });
            this.auth.accountId = accountId;
            return response.data;
        });
    }
    /**
     * Saves the user's session details.
     *
     * @see https://labs.ig.com/rest-trading-api-reference/service-detail?id=534
     */
    getSessionToken() {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = LoginAPI.URL.SESSION + '?fetchSessionTokens=true';
            const response = yield this.apiClient.get(resource);
            this.auth.securityToken = response.headers['x-security-token'];
            this.auth.clientSessionToken = response.headers.cst;
            return true;
        });
    }
    /**
     * Creates a session from predefined token values.
     */
    createSessionFromToken(securityToken, cst, accountId, lightstreamerEndpoint) {
        this.auth.accountId = accountId;
        this.auth.clientSessionToken = cst;
        this.auth.lightstreamerEndpoint = lightstreamerEndpoint;
        this.auth.securityToken = securityToken;
    }
    /**
     * Creates a session using the IG Mobile App API.
     *
     * WARNING: This endpoint only works with a production environment.
     */
    createSessionFromMobileLogin(username, password) {
        return __awaiter(this, void 0, void 0, function* () {
            delete this.auth.accessToken;
            const resource = 'https://api.ig.com/clientsecurity/session';
            const response = yield axios_1.default.post(resource, {
                enc: false,
                password,
                username: username,
            }, {
                'axios-retry': {
                    retries: 0,
                },
            });
            this.auth.securityToken = response.headers['x-security-token'];
            this.auth.clientSessionToken = response.headers.cst;
            this.auth.accountId = response.data.accountId;
            this.auth.lightstreamerEndpoint = response.data.lightstreamerEndpoint;
            return response.data;
        });
    }
    /**
     * Returns the user's session details.
     *
     * @see https://labs.ig.com/rest-trading-api-reference/service-detail?id=534
     */
    getSession() {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = LoginAPI.URL.SESSION + '?fetchSessionTokens=true';
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
    login(username, password) {
        return __awaiter(this, void 0, void 0, function* () {
            const isLive = this.apiClient.defaults.baseURL === APIClient_1.APIClient.URL_LIVE;
            if (isLive) {
                return this.createSessionFromMobileLogin(username, password);
            }
            return this.createSession(username, password);
        });
    }
    /**
     * Log out of the current session.
     *
     * @see https://labs.ig.com/rest-trading-api-reference/service-detail?id=600
     */
    logout() {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = LoginAPI.URL.SESSION;
            return this.apiClient.delete(resource, {
                'axios-retry': {
                    retries: 0,
                },
            });
        });
    }
    /**
     * Refreshes a trading session, obtaining new session tokens for subsequent API access.
     *
     * @see https://labs.ig.com/rest-trading-api-reference/service-detail?id=523
     */
    refreshToken() {
        return __awaiter(this, void 0, void 0, function* () {
            delete this.auth.accessToken;
            const resource = LoginAPI.URL.REFRESH_TOKEN;
            const response = yield this.apiClient.post(resource, {
                refresh_token: this.auth.refreshToken,
            });
            this.auth.accessToken = response.data.access_token;
            this.auth.refreshToken = response.data.refresh_token;
            return response.data;
        });
    }
}
exports.LoginAPI = LoginAPI;
LoginAPI.URL = {
    REFRESH_TOKEN: `/session/refresh-token`,
    SESSION: `/session`,
};
